### ISSUE TITLE
Consider replacing db impl with a native agnostic one

### ISSUE THREAD

**Issue by ericdallo** [clojure-lsp/clojure-lsp#703]
Datalevin works well for most cases but we have some headaches:

**pros**
- datalevin query and save is faster than our previous db impl (sqlite)
- datalevin has native parser for end which we use for analysis

**cons**
- Datalevin has some extra dependencies that doesn't make sense for lib usage. Related #658 
- Datalevin depends on native C libs for native image which bring some issues regarding OSs compatibility like aarch64
- some rare times clojure-lsp stuck on `finding cache`, which seems to be related with datalevin freezing
- We are already suppressing some warnings from lmdb-java regarding deprecated ByteArrayBuffer with some workarounds

**Comment by ericdallo** [comment]
I'll spend some time testing transit since it's the same clj-kondo uses for its cache and it works pretty well with native image. My biggest concerns are: reliability(what happens if clojure-lsp process is closed during a read or write of transit, do we have a chance to corrupt transit?) and performance

**Comment by ericdallo** [comment]
Made some tests with criterium, this is the time spent during read and write for clojure-lsp db case:
```
transit
Read 0.84 0.88 0.87 0.87 0.88 = 0.86
Write 2.33 2.15 2.30 2.90 2.43 = 2.42

datalevin
Read 2.88 2.05 2.11 2.11 2.20 = 2.27
Write 6.83 7.06 7.09 6.04 6.19 = 6.64
```

Overall, transit seems to fit clojure-lsp better, datalevin is a very good alternative but we are not and don't need to use all its potential and it's probably a overhead for what we seek as a cache db

**Comment by Cyrik** [comment]
It seems transit is already fast, but you might want to try https://github.com/ptaoussanis/nippy as well. It seems to be faster.
I'd stick with transit mostly because of the dependencies but nippy isn't too heavy either.

**Comment by ericdallo** [comment]
@Cyrik nippy is way faster indeed, almost 2 times than transit, but as you said have more dependencies, also I was thinking about making clojure-lsp db queryable by other tools or something like that in the future and transit may have a better standard for that I think, WDYT?
Anyway, nippy is a really good alternative as well, not tested graal support though but I guess it should work as well.

**Comment by Cyrik** [comment]
Yeah, I'd be very happy with transit, then I can easily consume it with omni-trace :) Just don't want to make LSP slower than it has to be. But I guess we'd be getting a 2 times speedup from switching away from datalevin and having more interop between tools is also a good thing to think about.

TLDR: I'm in camp transit 😄

**Comment by ericdallo** [comment]
@Cyrik yeah, As far as I tested, we will already improve a lot using transit, a dependency that is already available on classpath! thanks for the input!

**Comment by ericdallo** [comment]
After lots of tests, I merged transit to master! 🤞🏻
